File: Sources/BraintreeThreeDSecure/BTThreeDSecureResult.m
#import "BTThreeDSecureResult_Internal.h"

// MARK: - Swift File Imports for Package Managers
#if __has_include(<Braintree/Braintree-Swift.h>) // CocoaPods
#import <Braintree/Braintree-Swift.h>

#elif SWIFT_PACKAGE                              // SPM
/* Use @import for SPM support
 * See https://forums.swift.org/t/using-a-swift-package-in-a-mixed-swift-and-objective-c-project/27348
 */
@import BraintreeCore;
@import BraintreeCard;

#elif __has_include("Braintree-Swift.h")         // CocoaPods for ReactNative
/* Use quoted style when importing Swift headers for ReactNative support
 * See https://github.com/braintree/braintree_ios/issues/671
 */
#import "Braintree-Swift.h"

#else                                            // Carthage
#import <BraintreeCore/BraintreeCore-Swift.h>
#import <BraintreeCard/BraintreeCard-Swift.h>
#endif

// MARK: - Temporary Swift Module Imports
#if __has_include(<Braintree/BraintreeThreeDSecure.h>) // CocoaPods
#import <Braintree/Braintree-Swift.h>
#else                                            // SPM and Carthage
#import <BraintreeThreeDSecure/BraintreeThreeDSecure-Swift.h>
#endif

@implementation BTThreeDSecureResult

- (instancetype)initWithJSON:(BTJSON *)json {
    self = [super init];
    if (self) {
        if ([json[@"paymentMethod"] asDictionary]) {
            _tokenizedCard = [[BTCardNonce alloc] initWithJSON:json[@"paymentMethod"]];
        }

        if ([json[@"lookup"] asDictionary]) {
            _lookup = [[BTThreeDSecureLookup alloc] initWithJSON:json[@"lookup"]];
        }

        if ([json[@"errors"] asArray]) {
            NSDictionary *firstError = (NSDictionary *)[json[@"errors"] asArray].firstObject;
            if (firstError[@"message"]) {
                _errorMessage = [firstError[@"message"] asString];
            }
        } else {
            _errorMessage = [json[@"error"][@"message"] asString];
        }
    }
    return self;
}

@end

File: Sources/BraintreeThreeDSecure/BTThreeDSecureResult.swift
import Foundation

#if canImport(BraintreeCore)
import BraintreeCore
#endif

#if canImport(BraintreeCard)
import BraintreeCard
#endif

/// The result of a 3D Secure payment flow
@objcMembers public class BTThreeDSecureResult: NSObject {

    // MARK: - Public Properties

    /// The `BTCardNonce` resulting from the 3D Secure flow
    public var tokenizedCard: BTCardNonce?

    /// The result of a 3D Secure lookup. Contains liability shift and challenge information.
    public var lookup: BTThreeDSecureLookup?

    /// The error message when the 3D Secure flow is unsuccessful
    public var errorMessage: String?

    // MARK: - Initializer

    init(json: BTJSON) {
        if json["paymentMethod"].asDictionary() != nil {
            tokenizedCard = BTCardNonce(json: json["paymentMethod"])
        }

        if json["lookup"].asDictionary() != nil {
            lookup = BTThreeDSecureLookup(json: json["lookup"])
        }

        if let firstErrorMessage = json["errors"].asArray()?.first?["message"] {
            errorMessage = firstErrorMessage.asString()
        } else {
            errorMessage = json["error"]["message"].asString()
        }
    }
}
